File: public/PublicMedia/static/solutions/elixir/problem4.ex
# Purpose: Find the largest palindrome made from the product of two 3-digit numbers

# Author: Eben Quenneville
# Date: 2022-09-05, updated 2023-01-27
import ExUnit.Assertions

defmodule Problem4 do
  def palindrome?(n) do
    n == Integer.undigits(Enum.reverse(Integer.digits(n)))
  end

  def problem() do
    Enum.max(List.flatten(generate_palindromes()))
  end

  def generate_palindromes() do
    for x <- 100..1000 do
      for y <- x..1000 do
        if palindrome?(x * y) do
          x * y
        else
          0
        end
      end
    end
  end

  # Check solution
  def main() do
    solution = problem()
    assert solution == 906_609
    IO.puts("Output: #{solution}")
  end
end

Problem4.main()

File: public/PublicMedia/static/solutions/python/problem4.py
"""
Purpose: Find the largest palindrome made from the product of two 3-digit numbers

Author: Eben Quenneville
Date: 2022-09-04, modified 2023-01-29
"""


def problem4() -> int:
    largest = 9009
    for i in range(1000):
        for j in range(1000):
            result = i * j
            if is_palindrome(result) and result > largest:
                largest = result
    return largest


def main():
    solution = problem4()
    assert solution == 906609
    print(f"Output: {solution}")


def is_palindrome(num: int) -> bool:
    arr = [int(x) for x in str(num)]
    if arr == arr[::-1]:
        return True
    else:
        return False


if __name__ == "__main__":
    main()
